h/services/url_migration.py

```diff
@@ -0,0 +1,141 @@
+                # scheduled.
+                log.info("Skipping annotation %s", ann.uuid)
+                continue
+
+            ann_update_data = {
+                "target_uri": new_url_info["url"],
+            }
+
+            if "document" in new_url_info:
+                ann_update_data["document"] = transform_document(
+                    new_url_info["document"], new_url_info["url"]
+                )
+
+            # Add selectors to annotation if there is no selector of the same
+            # type.
+            #
+            # This change is specifically to aid in the migration of ebook
+            # annotations from a chapter/page URL to the containing book.
+            # The information about which chapter/page the annotation refers
+            # to is then moved into selectors.
+            #
+            # See https://github.com/hypothesis/h/issues/7709
+            if new_selectors := new_url_info.get("selectors"):
+                selectors = copy.deepcopy(ann.target_selectors)
+                for new_sel in new_selectors:
+                    if not any(s for s in selectors if s["type"] == new_sel["type"]):
+                        selectors.append(new_sel)
+                ann_update_data["target_selectors"] = selectors
+
+            # Update the annotation's `target_uri` and associated document,
+            # and create `Document*` entities for the new URL if they don't
+            # already exist.
+            h.storage.update_annotation(
+                self.request,
+                ann.id,
+                ann_update_data,
+                # Don't update "edited" timestamp on annotation cards.
+                update_timestamp=False,
+                reindex_tag="URLMigrationService.move_annotations",
+            )
+
+            log.info("Moved annotation %s to URL %s", ann.uuid, ann.target_uri)
+
+    def move_annotations_by_url(self, url, new_url_info):
+        """
+        Find annotations with a given target URL and move them to another URL.
+
+        :param url: The current URL of the annotations
+        :param new_url_info: The URL and document metadata of the new URL.
+             This is an entry matching the `URLMigrationSchema` schema.
+        """
+
+        session = self.request.db
+        uri_normalized = normalize(url)
+
+        # Get the IDs of all annotations on the old URL, and divide into
+        # fixed-sized batches. A separate Celery task is then launched per
+        # batch to migrate those annotations. This limits the amount of
+        # work per task in case there are a large number of annotations on
+        # a given URL.
+        annotations = (
+            session.query(Annotation)
+            .join(DocumentURI, Annotation.document_id == DocumentURI.document_id)
+            .filter(DocumentURI.uri_normalized == uri_normalized)
+            .options(load_only(Annotation.id, Annotation.target_uri))
+        )
+
+        ann_ids = [ann.id for ann in annotations]
```

Comment:
Since you only need the annotation IDs and not the full annotations, there might be a way to make the query more efficient by only retrieving the IDs in the first place. I can't remember if off the top of my head though

Reply:
> Since you only need the annotation IDs and not the full annotations, there might be a way to make the query more efficient by only retrieving the IDs in the first place.

This is what the `load_only` option above is intended to do.